tremor-pipeline/src/op/trickle/window.rs

```diff
@@ -487,6 +497,117 @@ impl Trait for No {
+                Ok(Actions::all_true())
+            } else {
+                Ok(Actions::all_false())
+            }
+        } else if let Some(r) = v
+            .as_array()
+            .and_then(|a| a.iter().map(Value::as_bool).collect::<Option<Vec<bool>>>())
+        {
+            if let [emit, include] = *r.as_slice() {
+                Ok(Actions { include, emit })
+            } else {
+                Err("A array return needs to be a two element array of booleans with the form `[emit, include]`".into())
+            }
+        } else {
+            Err(format!(
+                "Window returned `{}` this can't be used as a window reply",
+                v.encode()
+            )
+            .into())
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct TumblingOnState {
+    pub(crate) initial_state: Value<'static>,
+    pub(crate) max_groups: usize,
+    pub(crate) script: Script<'static>,
+    pub(crate) tick_script: Option<Script<'static>>,
+    pub(crate) state: Value<'static>,
+}
+
+impl TumblingOnState {
+    pub(crate) fn reset(&mut self) {
+        self.state = self.initial_state.clone();
+    }
+    pub fn from_stmt(
+        state: Value<'static>,
+        max_groups: usize,
+        script: Script<'static>,
+        tick_script: Option<Script<'static>>,
+    ) -> Self {
+        Self {
+            initial_state: state.clone(),
+            max_groups,
+            script,
+            state,
+            tick_script,
+        }
+    }
+}
+
+impl Trait for TumblingOnState {
+    fn max_groups(&self) -> usize {
+        self.max_groups
+    }
+    fn on_event(
+        &mut self,
+        data: &mut ValueAndMeta,
+        ingest_ns: u64,
+        origin_uri: &Option<EventOriginUri>,
+    ) -> Result<Actions> {
+        let context = EventContext::new(ingest_ns, origin_uri.as_ref());
+        let (unwind_event, event_meta) = data.parts_mut();
+        let value = stry!(self.script.run(
+            &context,
+            AggrType::Emit,
+            unwind_event,    // event
+            &mut self.state, // state for the window
+            event_meta,      // $
+        ));
+        match value {
+            Return::Emit { value, .. } => Actions::try_from(&value),
+            Return::EmitEvent { .. } => Actions::try_from(&*unwind_event),
+            Return::Drop { .. } => Err("State based window didn't provide a boolean return".into()),
+        }
+    }
+
+    fn on_tick(&mut self, ns: u64) -> Result<Actions> {
+        let context = EventContext::new(ns, None);
+        let mut unwind_event = Value::const_null();
+        let mut event_meta = Value::const_null();
+        // We protect against altering event by ensuring event and meta
+        // arent accessed during 'compile' time
+        if let Some(tick_script) = &self.tick_script {
+            let value = tick_script.run(
+                &context,
+                AggrType::Emit,
+                &mut unwind_event, // event
```

Comment:
would it make sense to have `event` be the tick nanoseconds?

Reply:
hm I was thinking about it but decided against it.

The tick script is triggered synthetically and not from any `event`. I found calling the data in here `event` will be misleading as it suggests there is an event triggering this. 

To emphasize this using `event` is completely forbidden in the trick script and will error.

Reply:
So if i want to have the current time to do some time based window, i need to call `system::nanotime()` everytime?

Reply:
yes, same as you'd do in the `script` section code can be copied between then this way :)

Reply:
It is sad, as we have the tick info already available. Would it make sense to have a function in the `std::win` module that just gets that number without asking the CPU for a time?

Reply:
Cool, that needs to be documented!

Reply:
done